[2.4.240] SHUTDOWN COMPACT failure

Hello,

In the version 2.4.240, we hit the following exception with a `SHUTDOWN COMPACT` command:

```
        at <my_class>
        at org.codehaus.mojo.exec.ExecJavaMojo.doMain(ExecJavaMojo.java:371)
        at org.codehaus.mojo.exec.ExecJavaMojo.doExec(ExecJavaMojo.java:360)
        at org.codehaus.mojo.exec.ExecJavaMojo.lambda$execute$0(ExecJavaMojo.java:280)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: General error: "java.util.concurrent.CompletionException: org.h2.mvstore.MVStoreException: java.lang.AssertionError [2.4.240/3]"; SQL statement:
SHUTDOWN COMPACT [50000-240]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:566)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        at org.h2.message.DbException.get(DbException.java:212)
        at org.h2.message.DbException.convert(DbException.java:407)
        at org.h2.command.Command.executeUpdate(Command.java:318)
        at org.h2.command.Command.executeUpdate(Command.java:250)
        at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:262)
        at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:231)
        at my.H2DatabaseManager.importH2Statements(H2DatabaseManager.java:379)
        ... 8 more
Caused by: java.util.concurrent.CompletionException: org.h2.mvstore.MVStoreException: java.lang.AssertionError [2.4.240/3]
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1807)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        ... 1 more
Caused by: org.h2.mvstore.MVStoreException: java.lang.AssertionError [2.4.240/3]
        at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
        at org.h2.mvstore.FileStore.serializeAndStore(FileStore.java:1446)
        at org.h2.mvstore.FileStore.lambda$storeIt$2(FileStore.java:1381)
        at org.h2.mvstore.FileStore.submitOrRun(FileStore.java:1406)
        at org.h2.mvstore.FileStore.storeIt(FileStore.java:1380)
        at org.h2.mvstore.MVStore.storeNow(MVStore.java:965)
        at org.h2.mvstore.MVStore.store(MVStore.java:930)
        at org.h2.mvstore.MVStore.commit(MVStore.java:900)
        at org.h2.mvstore.MVStore.commit(MVStore.java:892)
        at org.h2.mvstore.MVStore.registerUnsavedMemoryAndCommitIfNeeded(MVStore.java:1372)
        at org.h2.mvstore.MVMap.copy(MVMap.java:1200)
        at org.h2.mvstore.MVMap.copyFrom(MVMap.java:1176)
        at org.h2.mvstore.MVStore.lambda$compact$0(MVStore.java:435)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
        ... 3 more
Caused by: java.lang.AssertionError
        at org.h2.mvstore.Page$PageReference.clearPageReference(Page.java:1079)
        at org.h2.mvstore.Page$NonLeaf.releaseSavedPages(Page.java:1388)
        at org.h2.mvstore.FileStore.serializeAndStore(FileStore.java:1441)
        ... 15 more
```
Following the PR #4273, we set the system property `h2.compactThreads` to 1 and the error get away, so this is a simple workaround. If we enforce it to any other value than 1 (tested with 2 and 4), then this error rises.

Note that we are importing an H2 dump from a previous version with `RUNSCRIPT` and the latter works without any error. Also if we just call `SHUTDOWN` instead of `SHUTDOWN COMPACT`, this works also fine, so it seems to be a multithread issue.

Thanks for the work and best regards.

--- a/h2/src/main/org/h2/mvstore/Page.java
+++ b/h2/src/main/org/h2/mvstore/Page.java
@@ -1076,7 +1076,6 @@ public abstract class Page<K,V> implements Cloneable {
         void clearPageReference() {
             if (page != null) {
                 page.releaseSavedPages();
-                assert page.isSaved() || !page.isComplete();
                 if (page.isSaved()) {
                     assert pos == page.getPos();
                     assert count == page.getTotalCount() : count + " != " + page.getTotalCount();

--- a/h2/src/test/org/h2/test/bench/TestPerformance.java
+++ b/h2/src/test/org/h2/test/bench/TestPerformance.java
@@ -14,6 +14,7 @@ import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.text.NumberFormat;
 import java.util.ArrayList;
 import java.util.Properties;
 import org.h2.store.fs.FileUtils;
@@ -189,7 +190,7 @@ public class TestPerformance implements Database.DatabaseTest {
             System.out.println("Total time: " + db.getTotalTime() + " ms");
             int statPerSec = (int) (db.getExecutedStatements() * 1000L / db.getTotalTime());
             db.log("Statements per second", "#/s", statPerSec);
-            System.out.println("Statements per second: " + statPerSec);
+            System.out.println("Statements per second: " + NumberFormat.getInstance().format(statPerSec));
             System.out.println("GC overhead: " + (100 * db.getTotalGCTime() / db.getTotalTime()) + "%");
             collect = false;
             db.stopServer();

--- a/h2/src/test/org/h2/test/bench/TestScalability.java
+++ b/h2/src/test/org/h2/test/bench/TestScalability.java
@@ -14,6 +14,7 @@ import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.text.NumberFormat;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Properties;
@@ -105,7 +106,7 @@ public class TestScalability implements Database.DatabaseTest {
         }
 
         BenchB test = new BenchB() {
-            // Since we focus on scalability here, lets emphasize multi-threaded
+            // Since we focus on scalability here, lets emphasize multithreaded
             // part of the test (transactions) and minimize impact of the init.
             @Override
             protected int getTransactionsPerClient(int size) {
@@ -215,7 +216,7 @@ public class TestScalability implements Database.DatabaseTest {
                     int statPerSec = (int) (executedStatements * 1000L / totalTime);
                     db.log("Statements per second", "#/s", statPerSec);
                     collect = false;
-                    System.out.println("Statements per second: " + statPerSec);
+                    System.out.println("Statements per second: " + NumberFormat.getInstance().format(statPerSec));
                     System.out.println("GC overhead: " + (100 * totalGCTime / totalTime) + "%");
                     ArrayList<Measurement> measurements = db.reset();
                     runSequence.results.add(measurements);

--- a/h2/src/test/org/h2/test/store/TestDefrag.java
+++ b/h2/src/test/org/h2/test/store/TestDefrag.java
@@ -12,6 +12,7 @@ import java.sql.Connection;
 import java.sql.ResultSet;
 import java.sql.Statement;
 import java.text.NumberFormat;
+import java.util.concurrent.TimeUnit;
 
 import org.h2.test.TestBase;
 import org.h2.test.TestDb;
@@ -43,48 +44,56 @@ public class TestDefrag  extends TestDb {
         config.traceTest = true;
         try {
             config.cipher = null;
-            testIt();
+            testIt(860_000_000L);
             config.cipher = "AES";
-            testIt();
+            testIt(920_000_000L);
         } finally {
             config.cipher = cipher;
         }
     }
 
-    public void testIt() throws Exception {
+    public void testIt(long expectedSizeLimit) throws Exception {
         String dbName = getTestName();
         deleteDb(dbName);
         File dbFile = new File(getBaseDir(), dbName + SUFFIX_MV_FILE);
         NumberFormat nf = NumberFormat.getInstance();
+        long startNs;
         try (Connection c = getConnection(dbName)) {
             try (Statement st = c.createStatement()) {
-                st.execute("CREATE TABLE IF NOT EXISTS test (id INT PRIMARY KEY, txt varchar)" +
-                            " AS SELECT x, x || SPACE(200) FROM SYSTEM_RANGE(1,10000000)");
+                for (int i = 0; i < 4; i++) {
+                    st.execute("CREATE TABLE IF NOT EXISTS test" + i + " (id INT PRIMARY KEY, txt varchar)" +
+                                " AS SELECT x, x || SPACE(200) FROM SYSTEM_RANGE(1,10000000)");
+                }
                 st.execute("checkpoint");
             }
             long origSize = dbFile.length();
             String message = "before defrag: " + nf.format(origSize);
             trace(message);
-            assertTrue(message, origSize > 4_000_000_000L);
+            assertTrue(message, origSize > 11_000_000_000L);
+            startNs = System.nanoTime();
             try (Statement st = c.createStatement()) {
                 st.execute("shutdown defrag");
             }
         }
         long compactedSize = dbFile.length();
-        String message = "after defrag: " + nf.format(compactedSize);
-        assertTrue(message, compactedSize < 400_000_000L);
+        String message = "after defrag: " + nf.format(compactedSize) +
+                " time: " + TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs) + " ms";
+        assertTrue(message, compactedSize < expectedSizeLimit);
         trace(message);
 
         try (Connection c = getConnection(dbName + ";LAZY_QUERY_EXECUTION=1")) {
             try (Statement st = c.createStatement()) {
-                ResultSet rs = st.executeQuery("SELECT * FROM  test");
-                int count = 0;
-                while (rs.next()) {
-                    ++count;
-                    assertEquals(count, rs.getInt(1));
-                    assertTrue(rs.getString(2).startsWith(count + "   "));
+                for (int i = 0; i < 4; i++) {
+                    int count = 0;
+                    try (ResultSet rs = st.executeQuery("SELECT * FROM  test"+i)) {
+                        while (rs.next()) {
+                            ++count;
+                            assertEquals(count, rs.getInt(1));
+                            assertTrue(rs.getString(2).startsWith(count + "   "));
+                        }
+                    }
+                    assertEquals(10_000_000, count);
                 }
-                assertEquals(10_000_000, count);
             }
         }
         deleteDb(dbName);

